actionpack/lib/action_dispatch/middleware/server_timing.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+require "action_dispatch/middleware/server_timing_subscriber"
+
+module ActionDispatch
+  class ServerTiming
+    class << self
+      def add_timing(key:, dur: nil, desc: nil)
+        timings[key] = { dur: dur, desc: desc }
+      end
+
+      def to_header
+        headers = []
+        timings.each_pair do |key, data|
+          header = "#{key}"
+          header += ";desc=\"#{data[:desc]}\"" if data[:desc]
+          header += ";dur=#{data[:dur]}" if data[:dur]
+          headers << header
+        end
+        headers.join(", ")
+      end
+
+      def measure(key:, desc: nil, &blk)
+        ActiveSupport::Notifications.instrument("measure.server_timing", desc: desc, key: key) do
+          yield
+        end
+      end
+
+      def timings
+        Current.timings ||= {}
+      end
+    end
+
+    def initialize(app, default_events = [])
+      @app = app
+      subscribe_to_events(default_events) unless default_events.empty?
+    end
+
+    def call(env)
+      @app.call(env).tap do |_, headers, _|
+        add_timing_headers(headers)
+      end
+    end
+
+    private
+
+      def add_timing_headers(headers)
+        headers["Server-Timing"] = self.class.to_header
+      end
+
+      def subscribe_to_events(events)
+        if events.include?(:db_runtime)
+          ActiveSupport::Notifications.subscribe("process_action.action_controller") do |*args|
+            payload = args.extract_options!
+            ServerTiming.add_timing(
+              key: :database,
+              dur: payload[:db_runtime],
+              desc: "DB"
```

Comment:
These are what the default descriptions look like in the chrome dev tools.

![server_timing](https://user-images.githubusercontent.com/6386302/57658942-953be300-75ae-11e9-9ec4-ee900cfc41b8.png)

I tried to keep the descriptions minimal as per the spec recommendations to keep HTTP overhead small.

Reply:
I would much prefer `Database` over `DB`, since the savings is only 6 bytes but having an explicit English word is much better than having an abbreviation.

Reply:
good suggestion, updated!